pkg/driver/aws-ebs/aws_ebs_tags_controller.go

```diff
@@ -0,0 +1,483 @@
+	"k8s.io/klog/v2"
+
+	configv1 "github.com/openshift/api/config/v1"
+	operatorapi "github.com/openshift/api/operator/v1"
+	"github.com/openshift/csi-operator/pkg/clients"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+)
+
+const (
+	awsSecretNamespace     = "openshift-cluster-csi-drivers"
+	awsSecretName          = "ebs-cloud-credentials"
+	infrastructureResource = "cluster"
+	driverName             = "ebs.csi.aws.com"
+)
+
+type EBSVolumeTagsController struct {
+	name          string
+	commonClient  *clients.Clients
+	eventRecorder events.Recorder
+	failedQueue   workqueue.RateLimitingInterface
+}
+
+func NewEBSVolumeTagsController(
+	ctx context.Context,
+	name string,
+	commonClient *clients.Clients,
+	eventRecorder events.Recorder) factory.Controller {
+
+	c := &EBSVolumeTagsController{
+		name:          name,
+		commonClient:  commonClient,
+		eventRecorder: eventRecorder,
+		failedQueue:   workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+	// Define the informer for watching Infrastructure resource changes
+	infraInformer := c.commonClient.ConfigInformers.Config().V1().Infrastructures().Informer()
+
+	// Add event handler to process updates only when ResourceTags change
+	infraInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldInfra := oldObj.(*configv1.Infrastructure)
+			newInfra := newObj.(*configv1.Infrastructure)
+
+			// Compare old and new AWS ResourceTags
+			if !reflect.DeepEqual(oldInfra.Status.PlatformStatus.AWS.ResourceTags, newInfra.Status.PlatformStatus.AWS.ResourceTags) {
+				klog.Infof("AWS ResourceTags changed: triggering reconciliation")
+				syncCtx := factory.NewSyncContext(c.name, eventRecorder)
+				err := c.Sync(context.TODO(), syncCtx)
+				if err != nil {
+					klog.Errorf("Error syncing AWS Infrastructure: %v", err)
+					return
+				}
+			}
+		},
+	})
+
+	go c.startFailedQueueWorker(ctx)
+
+	return factory.New().WithSync(
+		c.Sync,
+	).ResyncEvery(
+		20*time.Minute,
+	).WithInformers(
+		infraInformer,
+	).ToController(
+		"EBSVolumeTagsController",
+		eventRecorder,
+	)
+}
+
+func (c *EBSVolumeTagsController) Sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	klog.V(4).Infof("EBSVolumeTagsController sync started")
+	defer klog.V(4).Infof("EBSVolumeTagsController sync finished")
+
+	opSpec, _, _, err := c.commonClient.OperatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorapi.Managed {
+		return nil
+	}
+
+	infra, err := c.getInfrastructure()
+	if err != nil {
+		return err
+	}
+	var infraRegion = ""
+	if infra.Status.PlatformStatus != nil && infra.Status.PlatformStatus.AWS != nil {
+		infraRegion = infra.Status.PlatformStatus.AWS.Region
+	}
+	ec2Client, err := c.getEC2Client(ctx, infraRegion)
+	if err != nil {
+		return err
+	}
+	err = c.processInfrastructure(infra, ec2Client)
```

Comment:
So, does this controller needs to be an opt-in or a default controller? I don't assume every OCP customer wants this feature and if tagging were to fail after OCP upgrade, their clusters will be degraded and we will have support nightmare.

Reply:
cc @jsafrane

Reply:
It is presented as enabled by default + no opt out for all HyperShift clusters in the [enhancement](https://github.com/openshift/enhancements/pull/1700).